### ISSUE TITLE
Extended ElementData Sync

### ISSUE THREAD

**Issue by Pirulax** [multitheftauto/mtasa-blue#638]
Trying a lot of different techniques to sync data between server and clients i'm always facing bandwidth and cpu drops on server-side or fps-drops on clientside.

Summarizing my experience, i would like to suggest a feature, something like

setElementDataEx(element, key, value, syncType=normal, syncTargets = root, forceSync=true)
forceElementDataSync()

This function should work in conjunction with the classic elementData, but the difference is how things will be synchronized.

syncType:
1)  only players listed in syncTargets will receive the update. if new player joins or reconnects the server, the date is hidden for him, but if same key and same value will be set targeting to this player, he should receive the update.

2) normal - direct update will receive only syncTargets, sets data as public, if new player join the server, he should receive this data as well.

3) stream - this data should be synchronized only when player streams the element or element already streamed in. syncTargets is ignored here, but can be used to filter players who should receive data when element streamed. When element streams out, the data should stay in memory, so next time only delta will be synced. (it's a little bit tricky, we need to store some timestamp of last sync, but it will save a bandwidth and cpu.)

forceSync:
if true - data should be directly sent like setElementData, if false - the MTA should push it into buffer and collect all changes until current scope ends or method forceElementDataSync() triggered. This means the data should be synchronized not like many single packets, but one batch operation, or can be splitted in to few buckets for better shipping and client-side processing. The order of operations should be keeped. If newly set value overlaps previously set one in pending buffer, it should be overwritten.

forceElementDataSync() - forces operations buffer to be synced.

p.s. i know that most of it can be done on lua side, but there is a few limitations with setElementData without sync, which forces you to keep a duplicated element data tree. Actually i already did most of it in lua, but some of operations causing FPS drops due to lua quickness, so i think it should be implemented closely to core :)

From [mantis](https://bugs.mtasa.com/view.php?id=9250).

**Comment by Einheit-101** [comment]
It would be more important to fix incredible CPU usage of setElementData() and include a feature that prevents multiple calls of setElementData with the same data. You can use 20x setElementData(root, "test", true) in 1 frame and the server (And all clients) will receive the same "true" element data 20 times while it would only be necessary to send 1 (final) element data with 1 "true" once the current Lua frame is finished.

**Comment by Pirulax** [comment]
Element datas should be stored in a buffer and synced by a specified interval(100ms or so, I use 1500ms in my Lua integration of this thing), that'd solve the problem you described.

**Comment by ccw808** [comment]
@Einheit-101 setElementData only sends the data if it has changed

**Comment by haron4igg** [comment]
@Pirulax oh thanks mate for moving it from mantis :+1:

**Comment by Pirulax** [comment]
Alright, so.
What I'll do is I'll simply add syncTo, adding buffer isn't necessary, since I don't think I'd make any sense(it only'd if you're stupid, and use elementData for a 3rd thing which it isn't intended to)

**Comment by qaisjp** [comment]
> 3) stream - this data should be synchronized only when player streams the element or element already streamed in. _syncTargets is ignored here, but can be used to filter players who should receive data when element streamed._ When element streams out, the data should stay in memory, so next time only delta will be synced. (it's a little bit tricky, we need to store some timestamp of last sync, but it will save a bandwidth and cpu.)

In [my original comment](https://bugs.mtasa.com/view.php?id=9250#c24715) I suggested the `type` and `targets` arguments should be merged. I don't think it should be merged anymore/

Also, this should just be added to the existing setElementData. The current `synchronize` argument (in `setElementData`) can be replaced with an argument `type` that accepts both booleans and enum strings.

I think triggerClientEvent would work fine to send data to just certain users, but I agree that triggering an event to just do setElementData is a little pointless. Is there a measurable difference in performance? If there isn't a measurable difference, is there any point doing exactly this for scripters?

```lua
addEvent("setClientElementData", true)
addEventHandler("setClientElementData", root, function(key, value)
    setElementData(source, key, value, false)
end)
```

What I do think, however, is that there might be some value in `type="stream"`. 

If there's value in `targets` alongside `type="stream"`, we may as well implement `targets` in the general case. Conversely, if `targets` is implemented in the general case, it must work with `type="stream"`.

---

I also mentioned that the `forceSync` arg should be removed, but that doesn't make sense as it implies that data is batched by default.

I think batching (forceSync) should be done as part of separate issue/feature/pull request. The new issue can include:
- the `forceSync=true` argument should be renamed to `batch=false`
- `forceElementDataSync()` renamed to `sendBatchedData()`
- discussion of whether or not `batch=true` should auto send batched data (maybe you want certain data to send immediately, and other data to be sent in one go?)
- discussion of feasibility, since you may as well just use a table instead of multiple keys/values
- discussion of replacing `batch=bool` with `batch=batchObject`, so that you can have multiple batches being composed in parallel? You don't really want to keep track of whether certain function calls will send of batches you are currently composing, and this would fix that problem.

**Comment by WADmitry** [comment]
I often need the following cases:

1. Synced data between server and only one client (one-side direction, from server to client)
2. Same, but two-side directions, from server to client and from client to server)
3. Single time set value from server to an one client (or group of targets) (unsynced)

So, something like
```
setElementDataExt( player, "key", value, sync=false, clientChanges=false )
setElementDataExt( { player, vehicle }, "key", value, sync=false, clientChanges=false )
```
where `sync` is "keep data synced" (of course it should send data only when it has been changed) and `clientChanges` is "accept chages made on client side".

I will be happy if this is supported.

**Comment by Pirulax** [comment]
Implementing several batches is a very good idea, kudos for you.

However, how would we implement `target` sync type?
Maybe by just sending out a dictionary, looping thru that dictionary and setting a bHasDataToBeSyncedInOnStream variable to true on every player in the packet?

Then onClientElementStream in check if the player has data to be synced in and then sync it it.
But I think it'd cause errors in scripts that try to get the value immediately after the element has streamed in, so we'd need to delay the event and only trigger it after we've retrieved the data. 

Replying to you WADmtry:
1. Can be implemented
2. Can be implemented as well, but i find it pointless, theres triggerServerEvent for that.
3. Same as with 2.

**Comment by WADmitry** [comment]
@Pirulax _Can be implemented as well, but i find it pointless, theres triggerServerEvent for that._
Sure. _All_ of these things can be implemented using events and some code (including all from the your issue). So I'm talking about simplier life for developers.

There are some points why I would prefer element data instead of event. For an example:
* It should automatically wait when the element on client-side will be ready and set the value after it. For example, I created an element (vehicle, player, etc) on the server side and exactly after it I define some element data. But for some reason (lags, etc) it is created on client side a few later than event was received. So if I use events then I have to complicate architecture (create element, then wait event from client about it ready, then set data on server side, trigger event and only then set data on client side) (actually this is exacly what I do now). Would be better if the engine do it (because MTA already do this for synced data, as I understand).
* Consistency. It's simpler to use just one function to set element data instead of setElementData on server + event + setElementData on client.

**Comment by ccw808** [comment]
> For example, I created an element (vehicle, player, etc) on the server side and exactly after it I define some element data. But for some reason (lags, etc) it is created on client side a few later than event was received

@WADmitry Please add a new issue with an example on how to reproduce this problem

**Comment by WADmitry** [comment]
@ccw808 It's not a problem but just normal (and correct) behaviour. For example, if you start a server and 500 players connected at the same time, you will have some events order (i.e. lag) (just an example, this may happen because of many reasons). Don't you ever see `ERROR: Server triggered clientside event (eventName), but event is not added clientside` although the event is defined? That is case.

**Comment by ccw808** [comment]
`ERROR: Server triggered clientside event (eventName), but event is not added clientside` should only occur when the client has not downloaded the resource yet.
When the resource is running on the client, create element + setElementData should never fail

**Comment by WADmitry** [comment]
> should only occur when the client has not downloaded the resource yet.

Yes. So, if player connected and I already created him and his vehicles on the server side, his client still may download resources and then `triggerClientEvent` will fail. It's thing I was talking about.

> When the resource is running on the client, create element + setElementData should never fail

Yep. It's the reason why I would prefer `setElementData` instead of `triggerClientEvent`

**Comment by Pirulax** [comment]
Could we please discuss the actual topic, so I can start coding it?
Would be good if we'd figure out a good way for that streaming-in based sync.
I've been thinking about it, but I didn't solve it. The only thing that came to my mind was to sync it based on a specific constant(!) range.
Constant range, because otherwise it'd be complicated to do it CPU usage friendly, and we'd waste way more resources on trying to process the range based sync data than the actual sync would use  up.

**Comment by Pirulax** [comment]
Okay, so current plan is:
```lua
bool setElementData ( element theElement, string key, var value, [bool synchronize = true / string syncType = previous value / "synced", element / table sendTo = previous value / root, queue = true, queueName = "default"] )
```

- synchronize 
**"synced"**(==true, ==1) - data will be visible to players that are under the element defined in sendTo, or are in the table.
**"unsynced"**(==false, ==0) - obvious
**previous**(==2 or nil) - the previous value

- sendTo  
**element** - if root or any other element except a team the data will be sent to very player under that element in the element tree, if a team element then it'll be send to the players in the team(Maybe discuss this, it may broke scripts).
**table** - a list that contains player elements
**"previous"**(or nil)  - The previously set value.

- queue(maybe call it batch?) - whenever the update packet should be queued with other packets and sent as a big packet every X ms(can be defined in mtasever.conf, default will be **50ms**)

- queueName the name of the queue the data should be in.


New function:
sendElementDataQueue([string queueName = "default"])
(maybe call it syncElementDataQueue?)

If a player joins all data will be sent to him that is visible to it's highest parent in the element tree(by default root). By "visible" I mean that sendTo is the player's parent element, or is one of the player's parent elements.

**Comment by qaisjp** [comment]
How do you propose to implement this feature?

- For targeting only certain players, is it easy enough to just send the packet to certain players?
- For "stream", can you share how you imagine/intend for data to flow? i.e. explain what happens when you call `setElementData`, and what happens when the player streams in an element. Where is data stored?

> [..] previous value [..]

Please can you elaborate? I think the same settings  should be resupplied every time to reduce confusion.

I think it is more natural (and safer) for the new arguments to determine how data is sent, and not modify "slot settings" (settings for keys).

**Comment by Pirulax** [comment]
Comment updated.
I was already editing it when you posted the comment, so sorry for that.

Edit:
>For "stream", can you share how you imagine/intend for data to flow? i.e. explain what happens when you call setElementData, and what happens when the player streams in an element. Where is data stored? 

I won't implement stream as its pretty hard to do it properly(and would waste more server and client resources that i'd gain).
Think about it. 
The data won't be instantly there, after you've synced the player in, which'd cause that you can't use element data when onClientElementStreamIn is triggered.
Constantly checking if a player will be in stream range and getting his data from the sever would be an option, but that'd eat client CPU, but... maybe this is a solution.


Edit 2:
> For targeting only certain players, is it easy enough to just send the packet to certain players?

Yes it's as easy to just send the packet to the targeted players.


> I think it is more natural (and safer) for the new arguments to determine how data is sent, and not modify "slot settings" (settings for keys).

Maybe, but I think this is more convenient, think about it...
You want to change the value of the data, but getting back the sendTo data is a pain in the ass(for example when you want to send something to a specific and partially constant group of player, police faction on a sever for example(note; a player may be in more factions at a time, so thats why you cant use a team element simply)

**Comment by Pirulax** [comment]
Any more comments?
Or I can do it the way Ive described?

**Comment by qaisjp** [comment]
Introducing "slot settings" is inconsistent and incompatible with the current API (where you must always provide `sync=false` if you do not want it to be synced).

It also introduces a weird bit of extra state, which makes it more confusing than convenient. I guarantee adding this feature will cause many scripters to send questions like "Why isn't my data being sent?" (answer: you need to reset the sendTo) or "Why is this data being sent?" (answer: you haven't done sendTo yet).

It's so much more straightforward to just be explicit every time you use `setElementData`.

There's also nothing you from implementing these "slot settings" in Lua yourself (but it obviously won't work cross-resource unless you use exports).

Please submit a separate feature proposal for batching. (This issue is pretty much a megathread of three separate improvements to `setElementData`.)

**Comment by Pirulax** [comment]
sry, bad number

**Comment by Pirulax** [comment]
#1055 kinda resolved this issue.
Should we keep it open?

Also, implementing a batch element data sender doesnt sound that hard, if it improves the performance I could take a look at it.

**Comment by ArranTuna** [comment]
Yes this request has been solved by the addition of element data subscription.
